vault/identity_store_util.go

```diff
@@ -0,0 +1,1244 @@
+		}
+	}
+
+	// Create a name if there isn't one already
+	if entity.Name == "" {
+		randomName, err := uuid.GenerateUUID()
+		if err != nil {
+			return fmt.Errorf("failed to generate entity name")
+		}
+		entity.Name = "entity-" + randomName
+	}
+
+	// Entity metadata should always be map[string]string
+	err = validateMetadata(entity.ExternalMetadata)
+	if err != nil {
+		return fmt.Errorf("invalid entity metadata: %v", err)
+	}
+
+	// Set the creation and last update times
+	if entity.CreationTime.IsZero() {
+		entity.CreationTime = time.Now()
+		entity.LastUpdateTime = entity.CreationTime
+	} else {
+		entity.LastUpdateTime = time.Now()
+	}
+
+	return nil
+}
+
+func (i *identityStore) deleteIdentitiesInEntityInTxn(txn *memdb.Txn, entity *entityStorageEntry, identities []*identityIndexEntry) error {
+	if entity == nil {
+		return fmt.Errorf("entity is nil")
+	}
+
+	if txn == nil {
+		return fmt.Errorf("txn is nil")
+	}
+
+	var remainList []*identityIndexEntry
+	var removeList []*identityIndexEntry
+
+	for _, item := range identities {
+		for _, identity := range entity.Identities {
+			if identity.ID == item.ID {
+				removeList = append(removeList, identity)
+			} else {
+				remainList = append(remainList, identity)
+			}
+		}
+	}
+
+	// Remove indentity indices from identities table for those that needs to
+	// be removed
+	for _, identity := range removeList {
+		identityToBeRemoved, err := i.memDBIdentityByIDInTxn(txn, identity.ID)
+		if err != nil {
+			return err
+		}
+		if identityToBeRemoved == nil {
+			return fmt.Errorf("identity was not indexed")
+		}
+		err = i.memDBDeleteIdentityInTxn(txn, identityToBeRemoved)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Update the entity with remaining items
+	entity.Identities = remainList
+
+	return nil
+}
+
+func (i *identityStore) deleteIdentityFromEntity(entity *entityStorageEntry, identity *identityIndexEntry) error {
+	if entity == nil {
+		return fmt.Errorf("entity is nil")
+	}
+
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	for identityIndex, item := range entity.Identities {
+		if item.ID == identity.ID {
+			entity.Identities = append(entity.Identities[:identityIndex], entity.Identities[identityIndex+1:]...)
+			break
+		}
+	}
+
+	return nil
+}
+
+func (i *identityStore) updateIdentityInEntity(entity *entityStorageEntry, identity *identityIndexEntry) error {
+	if entity == nil {
+		return fmt.Errorf("entity is nil")
+	}
+
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	identityFound := false
+	for identityIndex, item := range entity.Identities {
+		if item.ID == identity.ID {
+			identityFound = true
+			entity.Identities[identityIndex] = identity
+		}
+	}
+
+	if !identityFound {
+		return fmt.Errorf("identity does not exist in entity")
+	}
+
+	return nil
+}
+
+func (i *identityStore) memDBAssociateIdentitiesToEntityInTxn(txn *memdb.Txn, entity *entityStorageEntry, identities []*identityIndexEntry) error {
```

Comment:
Yeah, it is not used now, but it was in use at some i'th iteration of code. I had taken a little bit of pain to get it right, so I left it there thinking it might be of use. I think it'll have to go away :-)